crash_similarity.py

```diff
@@ -122,28 +124,64 @@ def train_model(corpus):
 
     return model
 
+# create distance_matrix using precalculate cosine distance from rwmd
+def create_distance_matrix(model, dictionary, docset, all_distances):
+    distances = np.zeros((len(dictionary), len(dictionary)), dtype=np.double)
+    for j, w in dictionary.items():
+        if w in docset:
+            distances[0:all_distances.shape[1], j] = all_distances[model.wv.vocab[w].index].transpose()
+
+    return distances
+
+
+#Code moodified from Gensim.keyedvector.py
+def wmdistance(model, words1, words2, all_distances):
+    dictionary = gensim.corpora.Dictionary(documents=[words1, words2])
+    vocab_len = len(dictionary)
+
+    # create bag of words from document
+    def create_bow(doc):
+        norm_bow = np.zeros(vocab_len, dtype=np.double)
+        bow = dictionary.doc2bow(doc)
+
+        for idx, count in bow:
+            norm_bow[idx] = count / float(len(doc))
+
+        return norm_bow
+
+    bow1 = create_bow(words1)
+    bow2 = create_bow(words2)
+
+    docset = set(words2)
+    distances = create_distance_matrix(model, dictionary, docset, all_distances)
+
+    return emd(bow1, bow2, distances)
+
 
 def top_similar_traces(model, corpus, stack_trace, top=10):
     model.init_sims(replace=True)
 
     similarities = []
 
     words_to_test = preprocess(stack_trace)
-    words_to_test_clean = [w for w in words_to_test if w in model]
+    words_to_test_clean = [w for w in np.unique(words_to_test).tolist() if w in model]
 
     # TODO: Test if a first sorting with the average vectors is useful.
     '''
     inferred_vector = model.infer_vector(words_to_test)
     sims = model.docvecs.most_similar([inferred_vector], topn=len(model.docvecs))
     '''
 
-    all_distances = 1 - np.dot(model.syn0norm, model.syn0norm[[model.vocab[word].index for word in words_to_test_clean]].transpose())
-    print(all_distances.shape)
+    # Cos-similarity
+    all_distances = np.array(1.0 - np.dot(model.wv.syn0norm, model.wv.syn0norm[[model.wv.vocab[word].index for word in words_to_test_clean]].transpose()), dtype=np.double)
+    print all_distances.shape
```

Comment:
This is not valid in Python 3, you need to wrap the argument in parentheses like this: `print(all_distances.shape)`.